jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java

```diff
@@ -392,83 +358,161 @@ public abstract File getFile()
      * list of resource names contained in the given resource.
      * Ordering is unspecified, so callers may wish to sort the return value to ensure deterministic behavior.
      *
      * @return a list of resource names contained in the given resource, or null.
      * Note: The resource names are not URL encoded.
+     * @deprecated Replace with {@link #getPath()} and {@link Files#walkFileTree(Path, FileVisitor)}.
      */
-    public abstract String[] list();
-
-    /**
-     * Returns the resource contained inside the current resource with the
-     * given name, which may or may not exist.
-     *
-     * @param path The path segment to add, which is not encoded.  The path may be non canonical, but if so then
-     * the resulting Resource will return true from {@link #isAlias()}.
-     * @return the Resource for the resolved path within this Resource, never null
-     * @throws IOException if unable to resolve the path
-     * @throws MalformedURLException if the resolution of the path fails because the input path parameter is malformed, or
-     * a relative path attempts to access above the root resource.
-     */
-    public abstract Resource addPath(String path)
-        throws IOException, MalformedURLException;
-
-    /**
-     * Get a resource from within this resource.
-     */
-    @Override
-    public Resource getResource(String path) throws IOException
+    @Deprecated(forRemoval = true)
+    public String[] list()
     {
-        return addPath(path);
-    }
+        try (DirectoryStream<Path> dir = Files.newDirectoryStream(getPath()))
+        {
+            List<String> entries = new ArrayList<>();
+            for (Path entry : dir)
+            {
+                String name = entry.getFileName().toString();
 
-    // FIXME: this appears to not be used
-    @SuppressWarnings("javadoc")
-    public Object getAssociate()
-    {
-        return _associate;
+                if (Files.isDirectory(entry))
+                {
+                    name += "/";
+                }
+
+                entries.add(name);
+            }
+            int size = entries.size();
+            return entries.toArray(new String[size]);
+        }
+        catch (DirectoryIteratorException e)
+        {
+            LOG.debug("Directory list failure", e);
+        }
+        catch (IOException e)
+        {
+            LOG.debug("Directory list access failure", e);
+        }
+        return null;
     }
 
-    // FIXME: this appear to not be used
-    @SuppressWarnings("javadoc")
-    public void setAssociate(Object o)
+    public Resource resolve(String subPath) throws IOException
     {
-        _associate = o;
+        // Check that the path is within the root,
+        // but use the original path to create the
+        // resource, to preserve aliasing.
+        if (URIUtil.canonicalPath(subPath) == null)
+            throw new MalformedURLException(subPath);
+
+        if (URIUtil.SLASH.equals(subPath))
+            return this;
```

Comment:
Stuff like this makes it very hard to correctly close resources, since you cannot not tell if the returned resource needs to be separately closed or not.

I'm wondering perhaps we should not make Resource closeable.  Instead rely on the sweep of the filesystem pool and perhaps some explicit flush operations when we stop/undeploy contexts?

We are so far from being well closed

Reply:
`Resource` feels like it's designed to be a lightweight object that purely is a pointer to the stuff-to-be-served (like java.nio.Path), but jar files got in the way, `Resource` was made `Closeable` with a no-op close in the most common case that's rarely called, even when it should be.

I've been thinking that `Resource` should be modified to be a lightweight object by making it not implement `Closeable` & requiring an explicit `Filesystem` creation before jar: URIs can be used. But I could not find an elegant way to do that and this would break a lot of functionality that assumes jar: resources can be created at any time.

It's another sizeable cleanup.

Reply:
I can see two ways out:
 * no single `Resource` impl ever implements `Closeable` and we have to explicitly 'mount' the jar file before it can be used
   * +1 makes jar mounting/unmounting lifecycle explicit
   * -1 API is a bit cumbersome and requires reviewing/modifying everywhere `Resource` is currently used
 * `Resource`s backed by jar files are implicitly 'mounted' and their impl class implements `Closeable`
   *  +1 API is more elegant and doesn't require code changes as jar mounting is implicit
   *  -1 we may forget to close implicitly mounted jars which would require reviewing/modifying everywhere `Resource` is currently used to fix, or a solid sweeping policy

option 1:
```java
        URI uri = URI.create("jar:file:/path/to/file.jar");

        // Resource does not implement Closeable anymore

        // using a jar:file URI throws if not explicitly 'mounted'
        assertThrows(Exception.class, () -> Resource.newResource(uri));

        // 'mount' the FS before using the jar:file URI then close the FS
        try (RefCountingFileSystem fs = PoolingPathResource.createFsIfNeeded(uri))
        {
            Resource resource = Resource.newResource(uri);
            // use resource
        }

        // 'mount' the FS before using the jar:file URI then use sweep() to close all the open ones
        PoolingPathResource.createFsIfNeeded(uri);
        Resource resource = Resource.newResource(uri);
        // use resource

        // delete 'mounted' jar file; won't work on Windows
        Files.delete(Path.of("/path/to/file.jar"));
        // close all FS of modified/deleted 'mounted' jar files still open
        PoolingPathResource.sweep();
```

option 2:
```java
        URI uri = URI.create("jar:file:/path/to/file.jar");

        // resources backed by file: or jrt: URIs are not Closeable
        assertThat(Resource.newResource(URI.create("file:/path/to/file.txt")) instanceof CloseableResource, is(false));

        // use a resource backed by jar:file: URIs that is 'auto-mounted';
        // such resources implement CloseableResource so close it immediately after use
        try (CloseableResource resource = (CloseableResource) Resource.newResource(uri))
        {
            // use resource
        }

        // use a resource backed by jar:file: URIs that is 'auto-mounted';
        // close it by sweeping modified/deleted 'mounted' jars
        Resource resource = Resource.newResource(uri);

        // delete 'mounted' jar file; won't work on Windows
        Files.delete(Path.of("/path/to/file.jar"));
        // close all FS of modified/deleted 'mounted' jar files still open
        PoolingPathResource.sweep();
```

Reply:
@lorban  we should consider that we are comming from:
 + option 0: All resources implement Closeable, but it is a noop in most of them, which is lucky because most usages never call it (and sometimes it is called on a shared resource when it should not be).  Caching is done behind the scenes in URL Handlers over which we have no control, amazingly it has not really been a problem... so we are probably closing just enough to make it OK.
 
I don't actually mind option 1, as I think we seldom get given a "jar:file:/var/lib/webapps/foobar.war!/WEB-INF/web.xml"  resource and exect to open it directly.   Typically we would be given "file:/var/lib/webapps/foobar.war" and we are the ones that decide that it is a usable war and thus set the resource base to "jar:file:/var/lib/webapps/foobar.war!/"... so part of that decision that it is a usable war file can be "can we pool a filesystem for it"

If it proves to be a pain, then there is always option 3: no Resource ever implements Closeable,  jar files are implicitly 'mounted' as needed and we rely on the sweeper or explicit eviction to not leak.  

Note that explicit eviction is not a big deal, since if the FS is still actually in use by a Resource, it will continue to be usable via its API.... I guess if somebody started again from the top level, then we could end up pooling it again?

So I think we should consider option 1 and check how much of a pain it actually will be.